### ISSUE TITLE
Stress Testing

### ISSUE THREAD

**Issue by patitonar** [poanetwork/token-bridge#51]
Here are some ideas about how to measure the performance of the bridge. These tests could also be used to test the reliability of the bridge (if it continues to work under load or if it crashes).

1. Make a script that generates transactions in one side of the bridge and checks how many blocks are generated in the other side until the transaction is reflected. Run this script multiple times and compute statistics of the waiting time measured in blocks.
Pseudo-code:
```
beginBlockNumber = getBlockNumber(foreign)
sendTransaction(home)
waitForTransaction(foreign)
endBlockNumber = getBlockNumber(foreign)
numberOfBlocks = endBlockNumber - beginBlockNumber
```

2. Generate transactions and check the state of the rabbit queues (for example, average message waiting time).
3. Analyze logs to get the elapsed times of each step (from the time an event is processed by a watcher until all validators sign it, from there until the transaction is generated in the other side of the chain, etc.)

**Comment by akolotov** [comment]
Thanks @patitonar for detailed description!
I have no questions for 2 and 3.
But for the item 1 we need to determine networks which are used for testing. E.g. if we use Ropsten, `numberOfBlocks` could be different from one run to another and it is OK since this test network will contain different number of concurrent transactions in different time periods.
That's why I would suggest to use two private chains for testing. Since they will not contain any other activities besides bridge transaction, `numberOfBlocks` could be approximately the same from one run to another. In this case we also should expect that the number of transactions included in every block is near to block gas limit. If so, the bridge completely fits in the network configuration.
What do you think?

**Comment by fvictorio** [comment]
I think that if we use two private chains without activity, then the number of blocks won't be a useful measure. Maybe it could be useful for measuring time, but I'm not sure about that either.

I agree that if we do the tests using testnets that have other activities, the results won't be reproducible, but I think the numbers we obtain may be more realistic. And we could do several runs and average them to take into account the other "random" activity.

**Comment by pablofullana** [comment]
On the one hand, I tend to think alike. I believe that measuring out in the wild will give us a more realistic sense of what end users could expect (and would experience).

On the other hand, there may be some situations is which we'd like to test in private networks. For example, when trying to replicate a known issue that may happen under certain circumstances (e.g. number of transactions). So, having this said, does it make any sense to make these target networks configurable?

**Comment by akolotov** [comment]
thanks @fvictorio! from my point of view it more preferable to see how fast the bridge in handling events and sending transactions. E.g. the following test could be done for this with one bridge instance:
* prepare one block on Home which is completely full with *'deposit'* transactions
* wait when the bridge will catch the corresponding events
The expectation is that the bridge sends *'signatures'* to Home as fast as all of them will be included in one block (actually i would expect that it will be two blocks) as well.
In other words bridge should be able to produce dozens of transactions in one second as so they will be mined in one block rather than distributed between dozens of blocks.

**Comment by fvictorio** [comment]
Maybe we should clarify some things so that we all are on the same page.

First is the fact that we have two possible things to measure here. One is if the
bridge crashes or otherwise stops working under heavy load. This is a boolean
result: either it does or it doesn't (or, alternatively, at which load does it
crashes?). The other is to measure some metric of performance.

Let's say that we are doing the latter. Then we should define this metric. Some
examples:

- Send a transaction and see how long does it take from the moment it is sent to
  one chain until it is observed in the other chain.
- The same, but measure blocks instead of timestamps (this was the metric
  proposed in my previous comment).
- How long does the transaction is "inside" the bridge (this, for example, could
  be done analyzing the logs).
- The average time a transaction is in the rabbit queue.

And there surely are other options.

Maybe we should think about this first and then discuss what is the best
environment to measure this metric. What do you think?

**Comment by akolotov** [comment]
@fvictorio thanks for raised questions.

When we are planning testing for blockchain environment to measure how fast information is handled we  need to take into account:
1. there are two different scenarios how information come to bridge:
  **1.1.** The bridge is looking for _events_ in one block - the recent block produced by a blockchain validator. This block will contain some expected number of transactions which depends on _gas limit_ parameter configured for that particular chain. Assuming that transactions which produce events handled by the bridge larger than usual balance changing transactions we definitely can say that one block will not contain more than 380 txs in POA Network. And since every block appear in some determined time - 5 seconds for POA Network, we could say that the bridge will get not more than 380 txs every 5 seconds. It means in order to allow bridge to function properly, we need to achieve performance as so at least 380 txs were handled in 5 seconds before the next block appears. Otherwise it will lead to situations when the bridge still processing _events_ when new bunch of transactions come so congestion will occur.
  **1.2.** The bridge is looking for events in bunch of blocks - all blocks mined from the last processed block. In this case we could have thousands of transactions accumulated in the chain. For sure, our desire is to achieve such kind of bridge performance to reduce time of handling these transactions in order to eliminate congestion. 

2. the way the bridge transactions delivered to the chain depends on:
  **2.1.** the network topology. Depending on performance of RPC node and its peers, the transaction could require relatively significant time before achieve a node which will schedule them to be included into a new block. The main point here is that the network topology is not fixed so time to reach mining node could be different for two consequential transactions. 
  **2.2.** _gas price_. Depending on traffic available at current time in the blockchain network and fees suggested by network participants for transactions validation, we cannot predict time (starting from the moment the bridge transaction reached a mining node and till the moment it is included in the block) which it will take to validate the bridge transaction. Even if we extract `fast` _gas price_ from the gas price oracle it guarantees nothing since the oracle provides a recommendation by using _gas price_ values in the past and it could be irrelevant for the current moment due to heavy traffic.
  **2.3** _network hashpower_. If the hashpower quite big it could lead uncles (blocks which will not be included in the chain) appearing. If the block is considered as the uncle all transactions in this block will be returned to the transaction pool (_txpool_) and they will be considered as unvalidated again. Especially it is applicable for Ethereum Foundation network. So, it is not possible to say whether the transaction is fixed in the blockchain before finalization time (actually it is the number of blocks included in the chain on top of the block with the targeted transaction) expires.   
  **2.4.** miners configuration (quite exotic). Some miners could be configured in such way to handle some transactions with highest priority even if _txpool_ contains transactions with higher_gas price_ Or the miner could skip some transactions intentionally so such transaction should wait for more 'unbiased' miner.

All items listed in no. 2 presented that it does not make sense from my point of view to measure time when a transaction produced by the bridge appeared in the blockchain.

But items listed in no. 1 shows that we really need to worry about:
* how fast an event is handled by the bridge. I am interesting in time between block(s) with corresponding events received by the bridge and the corresponding transaction is sent through RPC.
* delay between handling two consequential events
* delay between sending two consequential transactions
* number of transactions the bridge could receive at once from the blockchain without crash
* if transactions could be sent in parallel, number of transaction the bridge could send at once without crash.

Are you OK with such set of metrics? Do you think they needs to be extended with other set? Did I miss something?

**Comment by fvictorio** [comment]
Thanks @akolotov. I'm OK with those metrics. I think it's more than good enough for a first iteration; we can always improve the stress tests in the future.

> how fast an event is handled by the bridge. I am interesting in time between block(s) with corresponding events received by the bridge and the corresponding transaction is sent through RPC.
> delay between handling two consequential events;
> delay between sending two consequential transactions;

I think these three can be measured using the logs, provided we can identify the whole lifecycle of each event.

> number of transactions the bridge could receive at once from the blockchain without crash;
> if transactions could be sent in parallel, number of transaction the bridge could send at once without crash;
> limits (number of incoming transactions) when the bridge starts missing some events, e.g. was not able to send the corresponding transaction with signature.

For these other three, we should be able to generate arbitrarily large blocks with events that will be processed by the bridge. I'm not really sure how to do this. In testnets, we can't control the size of the block or which transactions will be mined. Maybe parity with instant seal can accumulate the received transactions in one block? Or parity with another setting.

Anyway, given a way to do this, we should send blocks with an exponentially increasing number of transactions and see when/if it crashes. We then can use the logs to see if the bottleneck is in the watchers or in the senders.

**Comment by akolotov** [comment]
@fvictorio 

> For these other three, we should be able to generate arbitrarily large blocks with events that will be processed by the bridge.

If you have no other transactions in the network the easiest way to get the block full of needed transactions is to send these hundreds of such transactions in one second. They will get in _txpool_ almost in the same time as so miners will try to include all these transactions in the next block.

Another way is to stop mining subsystem on the node, send transactions to the node and turn on the mining subsystem again. By doing this, the node will try to include all transactions in _txpool_ in the next block.

**Comment by akolotov** [comment]
Addressed
